bdf-parser/src/glyph.rs

```diff
@@ -1,130 +1,107 @@
+        let (input, encoding) = statement("ENCODING", parse_encoding)(input)?;
+        let (input, scalable_width) = opt(statement("SWIDTH", unsigned_xy))(input)?;
+        let (input, device_width) = opt(statement("DWIDTH", unsigned_xy))(input)?;
+        let (input, bounding_box) = statement("BBX", BoundingBox::parse)(input)?;
+        let (input, _) = multispace0(input)?;
+        let (input, bitmap) = parse_bitmap(input)?;
+
+        Ok((
+            input,
+            Self {
+                name,
+                encoding,
+                scalable_width,
+                device_width,
+                bounding_box,
+                bitmap,
+            },
+        ))
+    }
 }
 
-fn glyph_swidth(input: &[u8]) -> IResult<&[u8], Vec2> {
-    statement("SWIDTH", unsigned_xy)(input)
+fn parse_string(input: &str) -> IResult<&str, String> {
+    map(take_until_line_ending, String::from)(input)
 }
 
-fn glyph_bounding_box(input: &[u8]) -> IResult<&[u8], BoundingBox> {
-    statement("BBX", separated_pair(unsigned_xy, space1, signed_xy))(input)
+fn parse_encoding(input: &str) -> IResult<&str, Option<char>> {
+    map(parse_to_i32, |code| {
+        u32::try_from(code).ok().and_then(std::char::from_u32)
+    })(input)
 }
 
-fn glyph_bitmap(input: &[u8]) -> IResult<&[u8], Vec<u32>> {
-    let (input, _) = multispace0(input)?;
-    let (input, glyph_data) =
-        delimited(tag("BITMAP"), take_until("ENDCHAR"), tag("ENDCHAR"))(input)?;
-
-    Ok((
-        input,
-        glyph_data
-            .to_vec()
-            .iter()
-            .filter(|d| is_hex_digit(**d))
-            .collect::<Vec<&u8>>()
-            .chunks(8)
-            .map(|c| {
-                c.iter()
-                    .rev()
-                    .enumerate()
-                    .map(|(k, &&v)| {
-                        let digit = v as char;
-                        digit.to_digit(16).unwrap_or(0) << (k * 4)
-                    })
-                    .sum()
-            })
-            .collect(),
-    ))
+fn parse_bitmap(input: &str) -> IResult<&str, Vec<u8>> {
+    map_parser(
+        delimited(tag("BITMAP"), take_until("ENDCHAR"), tag("ENDCHAR")),
+        preceded(multispace0, many0(terminated(parse_hex_byte, multispace0))),
+    )(input)
 }
 
-pub fn glyph(input: &[u8]) -> IResult<&[u8], Glyph> {
-    let (input, name) = glyph_startchar(input)?;
-    let (input, charcode) = glyph_charcode(input)?;
-    let (input, _) = opt(glyph_swidth)(input)?;
-    let (input, _) = opt(glyph_dwidth)(input)?;
-    let (input, bounding_box) = glyph_bounding_box(input)?;
-    let (input, bitmap) = glyph_bitmap(input)?;
-
-    Ok((
-        input,
-        Glyph {
-            bitmap,
-            bounding_box,
-            charcode,
-            name,
-        },
-    ))
+fn parse_hex_byte(input: &str) -> IResult<&str, u8> {
+    map_res(take_while_m_n(2, 2, |c: char| c.is_ascii_hexdigit()), |v| {
+        u8::from_str_radix(v, 16)
+    })(input)
```

Comment:
Do you think this would be cleaner written as:

```suggestion
    map_res(take(2usize), |v| u8::from_str_radix(v, 16))(input)
```

Performance remains about the same in the single benchmark included in the repo and it looks a bit cleaner. Thoughts?

Reply:
That's much nicer! I didn't realize that `from_str_radix` already checks that the input is a valid hex string.